test/storage/metric_test.cpp

```diff
@@ -0,0 +1,516 @@
+    auto table = table_handle.GetTable(txn_, txn_metric_table_);
+
+    for (uint8_t j = 0; j < num_txns_; j++) {
+      auto txn_id = id_map[j];
+      std::vector<type::TransientValue> search_vec;
+      search_vec.emplace_back(
+          type::TransientValueFactory::GetBigInt(static_cast<int64_t>(static_cast<uint64_t>(txn_id))));
+      auto row = table->FindRow(txn_, search_vec);
+      auto latency = type::TransientValuePeeker::PeekBigInt(row[1]);
+      auto read_cnt = type::TransientValuePeeker::PeekBigInt(row[2]);
+      auto insert_cnt = type::TransientValuePeeker::PeekBigInt(row[3]);
+      auto delete_cnt = type::TransientValuePeeker::PeekBigInt(row[4]);
+      auto update_cnt = type::TransientValuePeeker::PeekBigInt(row[5]);
+      EXPECT_EQ(read_cnt, read_map[txn_id]);
+      EXPECT_EQ(update_cnt, update_map[txn_id]);
+      EXPECT_EQ(insert_cnt, insert_map[txn_id]);
+      EXPECT_EQ(delete_cnt, delete_map[txn_id]);
+      EXPECT_GE(latency_max_map[txn_id], latency);
+      EXPECT_LE(latency_min_map[txn_id], latency);
+    }
+  }
+}
+
+/**
+ *  Testing metric stats collection and persistence, multiple threads
+ */
+// NOLINTNEXTLINE
+TEST_F(MetricTests, MultiThreadTest) {
+  const uint32_t num_threads = MultiThreadTestUtil::HardwareConcurrency();
+  common::WorkerPool thread_pool(num_threads, {});
+
+  catalog::table_oid_t table_oid = static_cast<catalog::table_oid_t>(2);  // any value
+  const catalog::db_oid_t database_oid(catalog::DEFAULT_DATABASE_OID);
+  auto db_handle = catalog_->GetDatabaseHandle();
+  const catalog::namespace_oid_t namespace_oid =
+      db_handle.GetNamespaceHandle(txn_, database_oid).NameToOid(txn_, default_namespace_);
+
+  for (uint8_t i = 0; i < num_iterations_; i++) {
+    common::ConcurrentQueue<transaction::timestamp_t> txn_queue;
+    common::ConcurrentMap<transaction::timestamp_t, int64_t> latency_max_map;
+    common::ConcurrentMap<transaction::timestamp_t, int64_t> latency_min_map;
+    storage::metric::StatsAggregator aggregator(txn_manager_, catalog_, nullptr);
+    common::ConcurrentVector<storage::metric::ThreadLevelStatsCollector *> collectors;
+    auto num_read = static_cast<uint8_t>(std::uniform_int_distribution<uint8_t>(1, UINT8_MAX)(generator_));
+    auto num_update = static_cast<uint8_t>(std::uniform_int_distribution<uint8_t>(1, UINT8_MAX)(generator_));
+    auto num_insert = static_cast<uint8_t>(std::uniform_int_distribution<uint8_t>(1, UINT8_MAX)(generator_));
+    auto num_delete = static_cast<uint8_t>(std::uniform_int_distribution<uint8_t>(1, UINT8_MAX)(generator_));
+
+    auto workload = [&](uint32_t id) {
+      // NOTICE: thread level collector must be alive while aggregating
+      if (id == 0) {  // aggregator thread
+        std::this_thread::sleep_for(aggr_period_);
+        aggregator.Aggregate(txn_);
+        auto iter = collectors.Begin();
+        for (; iter != collectors.End(); iter++) {
```

Comment:
`auto iter = collectors.Begin();` can be a part of the for loop. It is not being used anywhere else.

Reply:
Changed